stl/inc/future

```diff
@@ -184,23 +184,39 @@ struct _State_deleter : _Deleter_base<_Ty> { // manage allocator and deletion st
     _Alloc _My_alloc;
 };
 
+template <class _Ty>
+union _Result_holder {
+    _Result_holder() {}
+    ~_Result_holder() {}
+
+    _Ty _Held_value;
+    operator _Ty&() noexcept {
+        return _Held_value;
+    }
+};
+
 template <class _Ty>
 class _Associated_state { // class for managing associated synchronous state
 public:
     using _State_type = _Ty;
     using _Mydel      = _Deleter_base<_Ty>;
 
+    // TRANSITION: _Associated_state ctor will try to default construct a _Ty
     _Associated_state(_Mydel* _Dp = nullptr)
         : _Refs(1), // non-atomic initialization
           _Exception(), _Retrieved(false), _Ready(false), _Ready_at_thread_exit(false), _Has_stored_result(false),
-          _Running(false), _Deleter(_Dp) {
-        // TRANSITION: _Associated_state ctor assumes _Ty is default constructible
-    }
+          _Running(false), _Deleter(_Dp) {}
 
     virtual ~_Associated_state() noexcept {
-        if (_Has_stored_result && !_Ready) { // registered for release at thread exit
+        if (_Already_has_stored_result() && !_Ready) { // registered for release at thread exit
             _Cond._Unregister(_Mtx);
         }
+
+        if constexpr (!is_default_constructible_v<_Ty>) {

```

Comment:
`is_default_constructible_v<const int>` is `true`, since `const int` can be value-initialized, but `default_initalizable<const int>` is `false` since `const int` cannot be default-initialized. Should we be using something like `default_initializable` (but obviously not a concept, since we need it in C++14 mode) to discriminate instead of `is_default_constructible_v`?

Reply:
(If we do make a change here, we should add a test case for `promise<const int>` / `future<const int>`.)

Reply:
After some discussion in Discord with @StephanTLavavej, we're happy to accept this as-is (without support for `const meow`) and we'll file a followup issue to make that improvement later. This is a _huge_ improvement over the status quo. I won't resolve this comment so we remember to file the followup, but feel free to ignore or implement it as you like.

Reply:
Filed #2599.